pkg/controller/node/cidr_allocator.go

```diff
@@ -17,144 +17,171 @@ limitations under the License.
+		go func(stopChan <-chan struct{}) {
+			for {
+				select {
+				case workItem, ok := <-ra.nodeCIDRUpdateChannel:
+					if !ok {
+						glog.Warning("NodeCIDRUpdateChannel read returned false.")
+						return
+					}
+					ra.updateCIDRAllocation(workItem)
+				case <-stopChan:
+					glog.V(0).Info("StopChannel is closed.")
+					return
+				}
+			}
+		}(wait.NeverStop)
 	}
-	r.nextCandidate = (nextUnused + 1) % r.maxCIDRs
-
-	r.used.SetBit(&r.used, nextUnused, 1)
-
-	j := uint32(nextUnused) << uint32(32-r.subNetMaskSize)
-	ipInt := (binary.BigEndian.Uint32(r.clusterIP)) | j
-	ip := make([]byte, 4)
-	binary.BigEndian.PutUint32(ip, ipInt)
 
-	return &net.IPNet{
-		IP:   ip,
-		Mask: net.CIDRMask(r.subNetMaskSize, 32),
-	}, nil
+	return ra
 }
 
-func (r *rangeAllocator) Release(cidr *net.IPNet) error {
-	used, err := r.getIndexForCIDR(cidr)
+// AllocateOrOccupyCIDR looks at the given node, assigns it a valid CIDR
+// if it doesn't currently have one or mark the CIDR as used if the node already have one.
+func (r *rangeAllocator) AllocateOrOccupyCIDR(node *api.Node) error {
+	if node.Spec.PodCIDR != "" {
+		_, podCIDR, err := net.ParseCIDR(node.Spec.PodCIDR)
+		if err != nil {
+			return fmt.Errorf("failed to parse node %s, CIDR %s", node.Name, node.Spec.PodCIDR)
+		}
+		if err := r.cidrs.occupy(podCIDR); err != nil {
+			return fmt.Errorf("failed to mark cidr as occupied: %v", err)
+		}
+		return nil
+	}
+	podCIDR, err := r.cidrs.allocateNext()
 	if err != nil {
-		return err
+		recordNodeStatusChange(r.recorder, node, "CIDRNotAvailable")
+		return fmt.Errorf("failed to allocate cidr: %v", err)
 	}
 
-	r.lock.Lock()
-	defer r.lock.Unlock()
-	r.used.SetBit(&r.used, used, 0)
-
+	glog.V(10).Infof("Putting node %s with CIDR %s into the work queue", node.Name, podCIDR)
+	r.nodeCIDRUpdateChannel <- nodeAndCIDR{
+		nodeName: node.Name,
+		cidr:     podCIDR,
+	}
 	return nil
 }
 
-func (r *rangeAllocator) MaxCIDRs() int {
-	return r.maxCIDRs
+// ReleaseCIDR released the CIDR of the removed node
+func (r *rangeAllocator) ReleaseCIDR(node *api.Node) error {
+	if node.Spec.PodCIDR == "" {
+		return nil
+	}
+	_, podCIDR, err := net.ParseCIDR(node.Spec.PodCIDR)
+	if err != nil {
+		return fmt.Errorf("Failed to parse node %s, CIDR %s", node.Name, node.Spec.PodCIDR)
+	}
+
+	glog.V(4).Infof("recycle node %s CIDR %s", node.Name, podCIDR)
+	if err = r.cidrs.release(podCIDR); err != nil {
+		return fmt.Errorf("Failed to release cidr: %v", err)
+	}
+	return err
 }
 
-func (r *rangeAllocator) Occupy(cidr *net.IPNet) (err error) {
-	begin, end := 0, r.maxCIDRs
-	cidrMask := cidr.Mask
-	maskSize, _ := cidrMask.Size()
+// Marks all CIDRs with subNetMaskSize that belongs to serviceCIDR as used, so that they won't be
+// assignable.
+func (r *rangeAllocator) filterOutServiceRange(serviceCIDR *net.IPNet) {
+	if !r.clusterCIDR.Contains(serviceCIDR.IP.Mask(r.clusterCIDR.Mask)) && !serviceCIDR.Contains(r.clusterCIDR.IP.Mask(serviceCIDR.Mask)) {
```

Comment:
What this if is supposed to check? Can you please add a comment? {I was looking into it for some time, and still don't really follow it...]